port/zos390/omrvmem.c

```diff
@@ -1296,24 +1298,31 @@ omrvmem_default_large_page_size_ex(struct OMRPortLibrary *portLibrary, uintptr_t
 	}
 #endif /* OMR_ENV_DATA64 */
 
-	while (0 != PPG_vmem_pageSize[index]) {
-		if (0 != (OMRPORT_VMEM_MEMORY_MODE_EXECUTE & mode)) {
-			if ((ONE_M == PPG_vmem_pageSize[index]) &&
-				(0 != (OMRPORT_VMEM_PAGE_FLAG_PAGEABLE & PPG_vmem_pageFlags[index]))
-			) {
-				break;
-			}
-#if defined(OMR_ENV_DATA64)
-		} else {
-			if ((ONE_M == PPG_vmem_pageSize[index]) &&
-				(0 != (OMRPORT_VMEM_PAGE_FLAG_FIXED & PPG_vmem_pageFlags[index]))
-			) {
-				break;
-			}
-#endif
+	/* Determine maximum pageable, and fixed large pages available. */
+	while (0 != PG_vmem_pageSize[index] && ONE_M >= PPG_vmem_pageSize[index]) {
+
+		if (0 != (OMRPORT_VMEM_PAGE_FLAG_PAGEABLE & PPG_vmem_pageFlags[index]) {
+			/* Pageable. Keep track of largest pageable size */
+			maxPageableIndex = index;
+		}
+		else if (0 != (OMRPORT_VMEM_PAGE_FLAG_FIXED & PPG_vmem_pageFlags[index])) {
+			/* Fixed. Keep track of largest fixed size */
+			maxFixedIndex = index;
 		}
 		index++;
 	}
+
+	/* When mode is OMRPORT_VMEM_MEMORY_EXECUTE we will only consider pagable large pages.
+	 * When flag is OMRPORT_VMEM_PAGE_FLAG_PAGEABLE_PREFERABLE, we prefer pageable unless fixed is larger.
+	 */
+	if (0 != (OMRPORT_VMEM_MEMORY_MODE_EXECUTE & mode) ||
+	   (OMRPORT_VMEM_PAGE_FLAG_PAGEABLE_PREFERABLE == *pageFlags && PPG_vmem_pageSize[maxPageableIndex] >= PPG_vmem_pageSize[maxFixedIndex])) {
```

Comment:
Please add extra brackets, do not rely on order of operations

Reply:
Done. https://github.com/eclipse/omr/pull/4762/commits/65e4d3c847cdd2c81d5d1cd7f68d17e4eaff1ccf